executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,449 @@
+		matchedOuterPtrs:  make([][]byte, 0, 8),
+	}
+	iw.finished.Store(false)
+	return iw
+}
+
+func (iw *indexHashJoinInnerWorker) run(ctx context.Context) {
+	var task *indexHashJoinTask
+	joinResult, ok := iw.getNewJoinResult(ctx)
+	if !ok {
+		return
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task, ok = <-iw.taskCh:
+		}
+		if !ok {
+			break
+		}
+		if task.err != nil {
+			joinResult.err = task.err
+			break
+		}
+		err := iw.handleTask(ctx, task, joinResult)
+		if err != nil {
+			joinResult.err = err
+			break
+		}
+	}
+	if joinResult.err != nil || (joinResult.chk != nil && joinResult.chk.NumRows() > 0) {
+		select {
+		case iw.resultCh <- joinResult:
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (iw *indexHashJoinInnerWorker) getNewJoinResult(ctx context.Context) (*indexHashJoinResult, bool) {
+	joinResult := &indexHashJoinResult{
+		src: iw.joinChkResourceCh,
+	}
+	ok := true
+	select {
+	case joinResult.chk, ok = <-iw.joinChkResourceCh:
+	case <-ctx.Done():
+		return nil, false
+	}
+	return joinResult, ok
+}
+
+func (iw *indexHashJoinInnerWorker) buildHashTableForOuterResult(ctx context.Context, task *indexHashJoinTask) {
+	var (
+		keyBuf = make([]byte, 0, 64)
+		valBuf = make([]byte, 8)
+		err    error
+	)
+	for rowIdx, numRows := 0, task.outerResult.NumRows(); rowIdx < numRows; rowIdx++ {
+		var hasNull bool
+		if task.outerMatch != nil && !task.outerMatch[rowIdx] {
+			continue
+		}
+		row := task.outerResult.GetRow(rowIdx)
+		keyColIdx := iw.outerCtx.keyCols
+		for _, i := range keyColIdx {
+			if row.IsNull(i) {
+				hasNull = true
+				break
+			}
+		}
+		if hasNull {
+			continue
+		}
+		keyBuf, err = codec.HashChunkRow(iw.ctx.GetSessionVars().StmtCtx, keyBuf[:0], row, iw.outerCtx.rowTypes, keyColIdx)
+		if err != nil {
+			iw.finished.Store(true)
+			result, ok := iw.getNewJoinResult(ctx)
+			if !ok {
+				return
+			}
+			result.err = err
+			select {
+			case iw.resultCh <- result:
+			case <-ctx.Done():
+			}
+		}
+		*(*int)(unsafe.Pointer(&valBuf[0])) = rowIdx
+		task.lookupMap.Put(keyBuf, valBuf)
+	}
+	return
+}
+
+func (iw *indexHashJoinInnerWorker) fetchInnerResults(ctx context.Context, task *lookUpJoinTask) error {
+	lookUpContents, err := iw.constructLookupContent(task)
+	if err != nil {
+		return err
+	}
+	lookUpContents = iw.sortAndDedupLookUpContents(lookUpContents)
+	if iw.finished.Load().(bool) {
+		return nil
+	}
+	return iw.innerWorker.fetchInnerResults(ctx, task, lookUpContents)
+}
+
+func (iw *indexHashJoinInnerWorker) handleHashJoinInnerWorkerPanic(r interface{}) {
+	if r != nil {
+		iw.resultCh <- &indexHashJoinResult{err: errors.Errorf("%v", r)}
+	}
+	iw.wg.Done()
+}
+
+func (iw *indexHashJoinInnerWorker) handleTask(ctx context.Context, task *indexHashJoinTask, joinResult *indexHashJoinResult) error {
+	iw.wg = &sync.WaitGroup{}
+	iw.wg.Add(1)
+	go util.WithRecovery(func() { iw.buildHashTableForOuterResult(ctx, task) }, iw.handleHashJoinInnerWorkerPanic)
```

Comment:
How about postponing building hash table after we get all inner rows, and choose the smaller table to build the hash table?

Reply:
We can leave it as further optimization and do it in another PR since this PR is big enough.
How about adding some comments for it?